pkg/controllers/cluster/taint_manager.go

```diff
@@ -0,0 +1,347 @@
+}
+
+// Start starts an asynchronous loop that handle evictions.
+func (tc *NoExecuteTaintManager) Start(ctx context.Context) error {
+	bindingEvictionWorkerOptions := util.Options{
+		Name:          "binding-eviction",
+		KeyFunc:       nil,
+		ReconcileFunc: tc.syncBindingEviction,
+	}
+	tc.bindingEvictionWorker = util.NewAsyncWorker(bindingEvictionWorkerOptions)
+	tc.bindingEvictionWorker.Run(tc.ConcurrentReconciles, ctx.Done())
+
+	clusterBindingEvictionWorkerOptions := util.Options{
+		Name:          "cluster-binding-eviction",
+		KeyFunc:       nil,
+		ReconcileFunc: tc.syncClusterBindingEviction,
+	}
+	tc.clusterBindingEvictionWorker = util.NewAsyncWorker(clusterBindingEvictionWorkerOptions)
+	tc.clusterBindingEvictionWorker.Run(tc.ConcurrentReconciles, ctx.Done())
+
+	<-ctx.Done()
+	return nil
+}
+
+func (tc *NoExecuteTaintManager) syncBindingEviction(key util.QueueKey) error {
+	fedKey, ok := key.(keys.FederatedKey)
+	if !ok {
+		klog.Errorf("Failed to sync binding eviction as invalid key: %v", key)
+		return fmt.Errorf("invalid key")
+	}
+	cluster := fedKey.Cluster
+	keyStr := fedKey.String()
+
+	binding := &workv1alpha2.ResourceBinding{}
+	if err := tc.Client.Get(context.TODO(), types.NamespacedName{Namespace: fedKey.Namespace, Name: fedKey.Name}, binding); err != nil {
+		// The resource no longer exist, in which case we stop processing.
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get binding %s: %v", fedKey.NamespaceKey(), err)
+	}
+
+	if !binding.DeletionTimestamp.IsZero() || !util.IsTargetClusterExist(binding.Spec.Clusters, cluster) {
+		return nil
+	}
+
+	needEviction, tolerationTime, err := tc.needEviction(cluster, binding.Annotations)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if binding needs eviction", "binding", klog.KObj(binding))
+		return err
+	}
+
+	// Case 1: Need eviction now.
+	// Case 2: need eviction after toleration time. If time is up, do eviction right now.
+	// Case 3: Tolerate forever, we do nothing.
+	evictNow := false
+	if needEviction || tolerationTime == 0 {
+		evictNow = true
+	} else if tolerationTime > 0 {
+		now := time.Now()
+		if startTime, exist := tc.evictionStartTimeMap.get(keyStr); exist {
+			if startTime.Add(tolerationTime).Before(now) {
+				evictNow = true
+			}
+		} else {
+			tc.evictionStartTimeMap.set(keyStr, now)
+			tc.bindingEvictionWorker.AddAfter(fedKey, tolerationTime)
+		}
+	}
+
+	if evictNow {
+		if result := util.RemoveTargetCluster(binding.Spec.Clusters, cluster); len(result) < len(binding.Spec.Clusters) {
+			// update final result to evict the target cluster
+			binding.Spec.Clusters = result
+			if err = tc.Update(context.TODO(), binding); err != nil {
+				klog.ErrorS(err, "Failed to update binding", "binding", klog.KObj(binding))
+				return err
+			}
+			tc.emitClusterEvictionEventForResourceBinding(binding, cluster)
+			tc.evictionStartTimeMap.delete(keyStr)
+		}
+	}
+	return nil
```

Comment:
Should call `tc.evictionStartTimeMap.delete(keyStr)` if don't need to evict? (the cluster recovered and the taint is removed)

Reply:
That's right, I have made some changes. If trigger time has not been reached, return first.